### PR TITLE
Added HTTP Basic Support

### PR DESCRIPTION
**Pull request by kav**
Nothing too fancy here, just added the ability to send auth via a Basic header rather than as username and password fields.

### CODE REVIEW

**Comment by vesse** [comment]
Hey,

Sorry for the delay. Basic auth support has been discussed already in vesse/passport-ldapauth#4 and vesse/passport-ldapauth#30 and I still have no plans to add more logic not related to LDAP. It should not be too hard to use Basic Auth together with this strategy with custom callback or separate middleware.

**Comment by kav** [comment]
I'm not sure I understand your concern. You currently have code that posts username and passwords as body or query variables. As well as code to handle differing field names. That's both non-standard HTTP and it's not related to LDAP.
 
I suppose it makes sense if you are using a login form but once the user has logged in you need some mechanism to track login. An auth header on requests is the standard practice to maintain statelessness here.

There is not an easy way to build another middleware that interfaces with yours without intercepting and moving login data to the body or query which is a very strange approach, the reverse would be more sensible; that's to say a middleware that moves auth from query or body params to the auth header and the LDAP library doing things in the standard fashion.

The alternative you propose in all the closed pull requests is simply that your library doesn't offer an ldap auth strategy for passport and people should roll their own. I must admit that's pretty confusing. What's the point of this library if not to allow passport to properly support ldap-auth? The mechanism the front end of the application uses to communicate with the auth library which then communicates with the auth source is most definitely the business of a passport auth library.

Perhaps you should rename this project passport-http-form-ldapauth since that is in fact what it is.

**Comment by vesse** [comment]
I did decide to ignore your aggressive and impolite comment, but...

I have not been interested in including Basic Auth support because it is not generic like reading given field from body (form data, json, whatnot). Like your implemetation - it can handle *only* the Basic auth-scheme. Furthermore, handling even this one auth-scheme can be done in multiple ways which leaves room for bugs. For example, your implementation was different than in [basic-auth](https://github.com/jshttp/basic-auth) or in [passport-http](https://github.com/jaredhanson/passport-http). Third reason is that usually HTTP Basic Auth users don't need the features this strategy provides (which are username/password lookup, taken from [passport-local](https://github.com/jaredhanson/passport-local), and flash messages) so they can easily use [passport-http](https://github.com/jaredhanson/passport-http) with [ldapauth-fork](https://github.com/vesse/node-ldapauth-fork):

```javascript
passport.use(new BasicStrategy(function(user, pass, cb) {
  var ldap = new LdapAuth(opts);
  ldap.authenticate(user, pass, function(err, user) {
    ldap.close(function() {});
    cb(err, user);
  })
}));
```

But, after giving this a little thought I have decided to add an option where user can provide their own, synchronos credentials lookup method that will receive the `req` object as argument. This will enable header parsing but is not limited to basic auth only, and it leaves the actual parsing (and possible bugs related to it) out of this strategy. When done (I'll try to do this today) you could just do

```javascript
var basicAuth = require('basic-auth');
var ldapOpts = {
  server : { ... },
  credentialsLookup : basicAuth
}
```

to parse HTTP Basic Auth credentials.

**TLDR; I will add possibility for the user to provide a function that parses the credentials from the request. No need to be rude.**

**Comment by kav** [comment]
Apologies for letting my frustration show. Finding I was the third person to not only bring this up but to PR it and have it ignored and then closed with a recommendation to use ldapauth-fork and passport-http directly was very frustrating. I appreciate you can do whatever you want but it seems like a clear user need that you repeatedly responded to with a message that read, to me, as; "go build it yourself".

You do bring up a good point that your library has a lot of functionality that folks may not need. I really appreciate the extended AD work you've done and would hate to have to give that up and re-build it all myself. It's extremely handy during the interactive login phase that I can give detailed context.

Thanks for reexamining a lookup extension point as I think that'll serve my needs, and I'd guess the other folks as well.

**Comment by vesse** [comment]
This is now fixed in [`passport-ldapauth v0.6.0`](https://github.com/vesse/passport-ldapauth/releases/tag/v0.6.0) with the `credentialsLookup` option. [Basic Auth test case](https://github.com/vesse/passport-ldapauth/blob/f9105d3ec693d5144303f602c88f3c3797ed5c53/test/strategy-test.js#L213-L225) passes when the function is provided.
